Stop video track when disabling video

havardholvik opened the pull request:
In order to make the camera light turn off, we need to stop the video track completely.

Tested like:

1. `yarn dev` -> connect to room through Storybook
2. Toggle your camera, verify the camera light goes off

havardholvik committed:
Stop video track when disabling video

In order to make the camera light turn off, we need to stop the video
track completely.

nandito commented on package.json:
```diff
@@ -82,7 +82,7 @@
     "dependencies": {
         "@swc/helpers": "^0.3.13",
-        "@whereby/jslib-media": "whereby/jslib-media.git#1.3.1",
+        "@whereby/jslib-media": "whereby/jslib-media.git#havard/pan-259-make-sure-we-turn-off-camera-light-when-disabling-video",
```
Don't forget to update this.

havardholvik replied:
Will update once the `jslib-media` PR is reviewed and merged

havardholvik replied:
Updated

nandito commented on src/lib/LocalMedia.ts:
```diff
@@ -72,31 +82,76 @@ export default class LocalMedia extends TypedLocalMediaEventTarget {
+
+                track.enabled = false;
+
+                if (SHOULD_STOP_TRACK) {
```
Why do we need this `SHOULD_STOP_TRACK=true` constant and therefore this always `true` condition?

havardholvik replied:
Im reworking this to be dependent on whether we have passed an existing stream or not. So if we are passed a stream, we never stop the track, only disable it.

havardholvik replied:
Fixed

nandito commented on src/lib/react/useRoomConnection.ts:
```diff
@@ -311,6 +319,18 @@ function reducer(state: RoomConnectionState, action: RoomConnectionEvents): Room
                 screenshares: state.screenshares.filter((ss) => !ss.isLocal),
             };
+        case "LOCAL_CAMERA_ENABLED":
+            if (!state.localParticipant) return state;
```
When can `state.localParticipant` be false? (Before joining the room I guess? And is it expected to not do anything when client clicks the toggle mic btn?)

havardholvik replied:
That is a good question. Probably never in practice, but according to our types it can be, so this is needed in order to not get a type error on the lines below

nandito commented on src/lib/LocalMedia.ts:
```diff
@@ -44,9 +44,19 @@ const TypedLocalMediaEventTarget = EventTarget as { new (): LocalMediaEventTarge
     public screenshareStream?: MediaStream;
 
+    // Camera state
+    private _cameraEnabled = false;
```
The tests seems to be fail because the cameras are off initially now. Not sure if because of this line or not, but opening the test app and joining a room I seem my local participant's camera off until I explicitly enable it.

(At least locally, but I assume it's the same on CI)

<img width="1505" alt="Screenshot 2023-10-24 at 14 31 12" src="https://github.com/whereby/browser-sdk/assets/1159931/b6d3d500-fdf2-4fde-b8b3-784adcb58602">

havardholvik replied:
I know why these are failing, fix incoming

havardholvik replied:
It is in the case of explicitly passed localStream, then we basically end up with the incorrect local state for cam / mic enabled

havardholvik replied:
Should be fixed now

nandito reviewed (approved):
🍰

havardholvik committed:
Install sdk-dependencies in test app

We have previously failed to install the latest dependencies of the sdk
when installing the sample app. This change here makes sure
`yarn installed` is run AFTER adding the local yalc dependency.
